packages/react-graphql/src/hooks/useQuery.ts

```diff
@@ -0,0 +1,202 @@
+
+import {AsyncQueryComponentType} from '@shopify/react-graphql';
+
+import {
+  QueryHookOptions,
+  QueryHookResult,
+  QueryDocumentResolving,
+} from './types';
+import {useApolloClient} from './ApolloContext';
+import useGraphQLDocument from './useGraphQLDocument';
+import {
+  getCachedObservableQuery,
+  invalidateCachedObservableQuery,
+} from './queryCache';
+import {compact, objectToKey, actHack} from './utilities';
+
+export default function useQuery<Data = any, Variables = OperationVariables>(
+  queryOrComponent: DocumentNode<Data, Variables>,
+  options?: QueryHookOptions<Data, Variables>,
+): QueryHookResult<Data, Variables>;
+
+export default function useQuery<
+  Data = any,
+  Variables = OperationVariables,
+  DeepPartial = {}
+>(
+  queryOrComponent: AsyncQueryComponentType<Data, Variables, DeepPartial>,
+  options?: QueryHookOptions<Data, Variables>,
+): QueryDocumentResolving<Data, Variables> | QueryHookResult<Data, Variables>;
+
+export default function useQuery<
+  Data = any,
+  Variables = OperationVariables,
+  DeepPartial = {}
+>(
+  queryOrComponent:
+    | DocumentNode<Data, Variables>
+    | AsyncQueryComponentType<Data, Variables, DeepPartial>,
+  options: QueryHookOptions<Data, Variables> = {},
+): QueryHookResult<Data, Variables> | QueryDocumentResolving<Data, Variables> {
+  const {
+    skip = false,
+    pollInterval,
+    notifyOnNetworkStatusChange = false,
+    client: overrideClient,
+    context,
+    variables,
+    fetchPolicy,
+    errorPolicy,
+  } = options;
+
+  const [responseId, setResponseId] = useState(0);
+  const client = useApolloClient(overrideClient);
+  const query = useGraphQLDocument(queryOrComponent);
+
+  const watchQueryOptions: WatchQueryOptions<Variables> = useMemo(
+    () =>
+      compact({
+        context,
+        errorPolicy,
+        fetchPolicy,
+        notifyOnNetworkStatusChange,
+        pollInterval,
+        query,
+        variables,
+      }),
+    [
+      query,
+      pollInterval,
+      notifyOnNetworkStatusChange,
+      context && objectToKey(context),
+      variables && objectToKey(variables),
+      fetchPolicy,
+      errorPolicy,
+    ],
+  );
+
+  const observableQuery = useMemo(
+    () => {
+      if (!query) {
+        return;
+      }
+
+      // eslint-disable-next-line consistent-return
+      return getCachedObservableQuery<Data, Variables>(
+        client,
+        watchQueryOptions,
+      );
+    },
+    [query, client, watchQueryOptions],
+  );
+
+  useEffect(
+    () => {
+      if (skip || !observableQuery) {
+        return;
+      }
+
+      const invalidateCurrentResult = () => {
+        // A hack to get rid React warnings during tests. The default
+        // implementation of `actHack` just invokes the callback immediately.
+        // In tests, it's replaced with `act` from react-testing-library.
+        // A better solution welcome.
+        actHack(() => {
+          setResponseId(x => x + 1);
```

Comment:
_carry over things that's not resolved yet but lost due to moving files_

https://github.com/Shopify/quilt/pull/642#discussion_r273721557
> Hmm seems like a bad idea, not sure why we don't just use a useRef for the current response ID instead of state?